### PR TITLE
provider/aws: Add aws_alb_listener_rule resource

### PR DESCRIPTION
**Pull request by jen20**
This commit adds the aws_alb_listener_rule resource along with acceptance tests and documentation.

Part of #8137.

### CODE REVIEW

File: builtin/providers/aws/resource_aws_alb_listener_rule.go [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+		return errwrap.Wrapf("Error creating ALB Listener Rule: {{err}}", err)
+	}
+
+	if len(resp.Rules) == 0 {
+		return errors.New("Error creating ALB Listener Rule: no rules returned in response")
+	}
+
+	d.SetId(*resp.Rules[0].RuleArn)
+
+	return nil
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_alb_listener_rule.go]: should call `resourceAwsAlbListenerRuleRead` func right?

File: builtin/providers/aws/resource_aws_alb_listener_rule.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+	}
+
+	actions := make([]interface{}, len(rule.Actions))
+	for i, action := range rule.Actions {
+		actionMap := make(map[string]interface{})
+		actionMap["target_group_arn"] = *action.TargetGroupArn
+		actionMap["type"] = *action.Type
+		actions[i] = actionMap
+	}
+	d.Set("action", actions)
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_alb_listener_rule.go]: do we want to error check that the Set doesn't cause any issues here?

File: builtin/providers/aws/resource_aws_alb_listener_rule.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+		conditionMap := make(map[string]interface{})
+		conditionMap["field"] = *condition.Field
+		conditionValues := make([]string, len(condition.Values))
+		for k, value := range condition.Values {
+			conditionValues[k] = *value
+		}
+		conditionMap["values"] = conditionValues
+		conditions[i] = conditionMap
+	}
+	d.Set("condition", conditions)
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_alb_listener_rule.go]: do we want to error check that the Set doesn't cause any issues here?

File: builtin/providers/aws/resource_aws_alb_listener_rule.go [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+	resp, err := elbconn.ModifyRule(params)
+	if err != nil {
+		return errwrap.Wrapf("Error modifying ALB Listener Rule: {{err}}", err)
+	}
+
+	if len(resp.Rules) == 0 {
+		return errors.New("Error modifying creating ALB Listener Rule: no rules returned in response")
+	}
+
+	return resourceAwsAlbListenerRuleRead(d, meta)
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_alb_listener_rule.go]: the Read is calling the read - this should be `return nil`

Comment (jen20) [inline builtin/providers/aws/resource_aws_alb_listener_rule.go]: This is in the create?

File: builtin/providers/aws/resource_aws_alb_listener_rule.go [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsAlbListenerRule() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsAlbListenerRuleCreate,
+		Read:   resourceAwsAlbListenerRuleRead,
+		Update: resourceAwsAlbListenerRuleUpdate,
+		Delete: resourceAwsAlbListenerRuleDelete,
+		Importer: &schema.ResourceImporter{
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_alb_listener_rule.go]: I think it'd be good to follow up with an Import test :)

**Comment by stack72** [comment]
Tests pass :)

```
% make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSALBListenerRule_'
==> Checking that code complies with gofmt requirements...
/Users/stacko/Code/go/bin/stringer
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/08/19 13:12:46 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSALBListenerRule_ -timeout 120m
=== RUN   TestAccAWSALBListenerRule_basic
--- PASS: TestAccAWSALBListenerRule_basic (79.20s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	79.219s
```

**Comment by stack72** [comment]
This LGTM! We can follow up with the nitpicks post merge :)

File: builtin/providers/aws/resource_aws_alb_listener_rule.go (hunk 2 above)
Comment (jen20) [inline builtin/providers/aws/resource_aws_alb_listener_rule.go]: This is somewhat following convention - I'm not too worried about it here. In general though I'd agree we should formalise more of the policy about where we check errors on `d.Set()`

Comment (catsby) [inline builtin/providers/aws/resource_aws_alb_listener_rule.go]: I've formalized around checking the errors if `d.Set` is setting anything other than a scalar value

**Pull request merged**